webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,739 @@
+    } catch (err) {
+      Cu.reportError(err);
+      return null;
+    }
+    return providerDownloadFolder;
+  },
+
+  async handleMove(download) {
+    const providerDownloadFolder = await this.checkProviderDownloadFolder();
+    if (download.succeeded) {
+      await this._moveDownload(download, providerDownloadFolder);
+    }
+  },
+
+  async handleLocalMove(download) {
+    const downloadsDir = await Downloads.getPreferredDownloadsDirectory();
+    const downloadsDirExists = await this.checkIfAssetExists(downloadsDir);
+    if (downloadsDirExists) {
+      if (download.succeeded) {
+        await this._moveDownload(download, downloadsDir);
+      }
+    }
+  },
+};
+
+
+var WindowListener = {
+  setupBrowserUI: function wm_setupBrowserUI(window) {
+    const utils = window.QueryInterface(Ci.nsIInterfaceRequestor).getInterface(Ci.nsIDOMWindowUtils);
+    utils.loadSheetUsingURIString(CloudDownloadsView.stylesURL, Ci.nsIDOMWindowUtils.AUTHOR_SHEET);
+  },
+
+  tearDownBrowserUI: function wm_tearDownBrowserUI(window) {
+    const utils = window.QueryInterface(Ci.nsIInterfaceRequestor).getInterface(Ci.nsIDOMWindowUtils);
+    utils.removeSheetUsingURIString(CloudDownloadsView.stylesURL, Ci.nsIDOMWindowUtils.AUTHOR_SHEET);
+  },
+
+  // nsIWindowMediatorListener functions
+  onOpenWindow: function wm_onOpenWindow(xulWindow) {
+    // A new window has opened
+    const domWindow = xulWindow.QueryInterface(Ci.nsIInterfaceRequestor)
+      .getInterface(Ci.nsIDOMWindow);
+
+    // Wait for it to finish loading
+    domWindow.addEventListener("load", async function listener() {
+      // If this is a browser window or places library window then setup its UI
+      if (domWindow.document.documentElement.getAttribute("windowtype") === "navigator:browser" ||
+          domWindow.document.documentElement.getAttribute("windowtype") === "Places:Organizer" ) {
+        if (!CloudDownloadsView.stylesURL) {
+          return;
+        }
+        WindowListener.setupBrowserUI(domWindow);
+
+        const moveDownloadMenuItem = domWindow.document.getElementById("moveDownload");
+        if (moveDownloadMenuItem) {
+          return;
+        }
+        await CloudDownloadsView.registerContextMenu(domWindow);
+      }
+    }, {once: true});
+  },
+};
+
+/**
+ * generic pref that shows if cloud storage API is in use, by default set to false.
+ */
+XPCOMUtils.defineLazyPreferenceGetter(CloudDownloadsView, "gIsAPIEnabled",
+  CLOUD_SERVICES_PREF + "api.enabled", false, () => CloudDownloadsView.toggleAPIEnabledState());
+
+XPCOMUtils.defineLazyPreferenceGetter(CloudDownloadsInternal, "downloadDirSetting",
+  "browser.download.dir", "", () => CloudDownloadsInternal._setDefaultDownloadDirIconURL());
+
+XPCOMUtils.defineLazyPreferenceGetter(CloudDownloadsInternal, "preferredProviderKey",
+  CLOUD_SERVICES_PREF + "storage.key", "");
+
+CloudDownloadsView.promiseInit = CloudDownloadsView.toggleAPIEnabledState();
```

Comment:
`promiseInit` seems unused?

Reply:
More generally, why do we need to call `toggleAPIEnabledState()` here? Won't it be called via the webextension API's init() method? And until that happens, `.stylesURL` is not defined and we do (almost) nothing.

Reply:
Can remove promiseInit.

toggleAPIEnabledState is called from init only if API is enabled (which is subsequent restart after install case).

On first install or enable of extension, we are  setting stylesURL and than creating 'cloud.services.api.enabled.pref'  for first time and need above toggleAPIEnabledState call when CloudDownloadsView API loads.

https://github.com/mozilla/shield-cloudstorage/blob/cloud-storage-web-ext/webextension/privileged/ui/cloudstorage.js#L49